recommender/engine/rl_agent/action_selector.py

```diff
@@ -18,24 +50,64 @@ def __call__(self, weights: torch.Tensor, user: User, search_data: SearchData) -
             weights: Weights returned by an actor model, tensor of shape [K, SE], where:
                 - K is the number of services required for the recommendation
                 - SE is a service content tensor embedding dim
-            user: User for whom the recommendation is generated. User's accessed_services is used for narrowing the itemspace down.
+            user: User for whom the recommendation is generated.
+                User's accessed_services is used for narrowing the itemspace down.
             search_data: Information used for narrowing the itemspace down
 
         Returns:
             The tuple of recommended services ids
         """
 
-        # TODO: implement (performance is crucial)
-        pass
+        search_data_indices = self._get_service_indices(
+            retrieve_matching_service_ids(search_data)
+        )
+        user_accessed_indices = self._get_service_indices(
+            list(map(lambda s: s.id, user.accessed_services))
+        )
+
+        mask = torch.zeros(self.I)
+        mask[search_data_indices] = 1
+        mask[user_accessed_indices] = torch.rand(()).item()
+        mask[self.forbidden_indices] = -1
+
+        engagement_values = F.softmax(weights @ self.itemspace.T, dim=1)
+        recommended_indices = self._choose_recommended_indices(engagement_values, mask)
 
-    def _create_itemspace(self) -> torch.Tensor:
+        return self.index_id_map.iloc[recommended_indices].id.values.tolist()
+
+    def _choose_recommended_indices(self, engagement_values, mask):
+        masked_engagement_values = mask * engagement_values
+        top_K_indices = (
+            torch.topk(masked_engagement_values, self.K).indices.numpy().tolist()
+        )
+        indices = [top_K_indices[0][0]]
+
+        for k in range(1, self.K):
+            indices += [next(filter(lambda x: x not in indices, top_K_indices[k]))]
+
+        return indices
+
+    def _get_service_indices(self, ids):
+        return self.index_id_map[self.index_id_map.id.isin(ids)].index.values
+
+    def _create_itemspace(self, services_embedder) -> torch.Tensor:
         """
         Creates itemspace tensor.
 
         Returns:
-            itemspace: tensor of shape [SE, I] where: SE is service content tensor
+            itemspace: tensor of shape [I, SE] where: SE is service content tensor
              embedding dim and I is the number of services
         """
 
-        # TODO: implement
-        pass
+        all_services = Service.objects.order_by("id")
+        self.I = len(all_services)
+        self.index_id_map = pd.DataFrame(
+            sorted(all_services.distinct("id")), columns=["id"]
+        )
+
+        service_one_hot_tensors = torch.Tensor([s.tensor for s in all_services])
```

Comment:
Maybe it would be more clear to use torch.stack method.

Reply:
this method will be purged anyway during refactoring